org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/rename/LSPRenameHandler.java

```diff
@@ -97,19 +95,21 @@ public void setEnabled(Object evaluationContext) {
 			IDocument document = LSPEclipseUtils.getDocument((ITextEditor) part);
 			if (document != null && isEnable) {
 				try {
-					isEnable = !LanguageServiceAccessor.getLanguageServers(document, LSPRenameHandler::isRenameProvider)
-							.get(50, TimeUnit.MILLISECONDS).isEmpty();
+					// See if there is a server that supports rename. We don't use LanguageServers.anyMatching in this
+					// case because we want to time out and not block the UI if we have to wait for servers to start
+					// but we also want to be able to detect the timeout and schedule a handler update for when
+					// the definitive info on server capabilities is available
+					isEnable = LanguageServers.forDocument(document).withFilter(LSPRenameHandler::isRenameProvider)
```

Comment:
Why do we rewrite this code? Does it have an advantage over the previous one? If not, and I would not know of any, I would leave the code as it was.

Reply:
I was going to use your anyMatching(timeout, unit) overload, but then I saw you ended up not submitting that.

I know we are rewriting things in the new API so that they get the benefit of the concurrency guarantees (which are not needed here), but the medium-term goal is that LanguageServiceAccessor becomes internal only and that the many confusing overloads get deprecated and removed or made package-private.

Since we want to deprecate direct access to these low-level methods, and given I was in the class anyway, I thought I might as well port this method so that we have this area of the code now completely decoupled from the old API.

Reply:
Okay, I have looked at the code a bit more, and I understand the use case, it is a good idea to offer a consistent way to deal with that, but it is a separate issue that the request ordering. The pattern of "look for matches and if timeout enable on the background" seems useful but would apply to all the handlers (for example org.eclipse.lsp4e.operations.format.LSPFormatHandler.isEnabled()). There are more in the code base.
I would be for taking it out of this PR and creating a new PR for that, so that we can merge the rest or if we do it here, have something that we can apply on the other handlers as well, in the second case, what about adding a Consumer to `anyMatching` that will be called if timeout before returning `false`? In my opinion that would be the pattern to follow, but just to do one bit at a time, I would still put that in a different PR

Reply:
OK if you want I will back this change out